### DISCUSSION TITLE
Setup Storybook add-ons

### DISCUSSION THREAD

**Discussion by eddybrando** [Qiskit/qiskit.org#2383] [Ideas]
With our basic Storybook setup (#1143) we already include Storybook [essential add-ons](https://storybook.js.org/docs/vue/essentials/introduction):

- Docs
- Controls
- Actions
- Viewport
- Backgrounds
- Toolbars & globals

There are [many more official and community add-ons](https://storybook.js.org/addons/) in the Storybook ecosystem.

As part of https://github.com/Qiskit/qiskit.org/issues/884#issuecomment-733034345, we should define which additional add-ons should be included in our project, and then set these up.

Here are my suggestions.

### Official add-ons (besides those included as essential add-ons)

| Add-on | Should be included? | Comment |
|-|-|-|
| [Source](https://github.com/storybookjs/storybook/tree/master/addons/storysource) | No | It displays the story's code, not the component's, so it's not very useful. The component's code is already shown with the docs add-on.  | 
| [Storyshots](https://github.com/storybookjs/storybook/tree/master/addons/storyshots) | Yes | **Pros:** Reusable code between stories and snapshot test definitions. // **Cons:** Dependency on Storybook for snapshot testing. | 
| [Accessibility](https://github.com/storybookjs/storybook/tree/master/addons/a11y) | Yes | Very helpful hints to cover accessibility basics. | 
| [Console](https://github.com/storybookjs/storybook-addon-console) | Yes | Could help remove some noise logs from the browser native console related to WebSockets. | 
| [Links](https://github.com/storybookjs/storybook/tree/master/addons/links) | No | I can't think of a scenario where this could be applied in our project. But please correct me if I'm missing any. |

### Cherry-picked community add-ons (for the full list go to https://storybook.js.org/addons/)

| Add-on | Should be included? | Comment |
|-|-|-|
| [Paddings](https://github.com/rbardini/storybook-addon-paddings) | Yes | Could be useful when testing shadows. |
| [Screenshot](https://github.com/reg-viz/storycap) | Yes | Useful UI assets ready to be shared. |
| [Open Component Code](https://github.com/product-ride/storybook-addon-vscode-source) | No | **Pros:** Handy shortcuts. **Cons:** It only works only for VSCode, making it not inclusive. We shouldn't restrict the usage of such a feature to only one code editor. |
| [Figma](https://github.com/pocka/storybook-addon-designs) | Yes | Helpful to cross-check design implementations.  |

---

Part of https://github.com/Qiskit/qiskit.org/issues/884#issuecomment-733034345

**Comment by y4izus** [comment]
Almost agree with everything. 

In my opinion, our project isn't too big and hasn't too many complicated components, so I think `Console` and `Paddings` aren't necessary right now.

`Storyshots` and `Screenshot` are both for taking snapshots, what differences do you find and why do you think we should add both?

**Comment by eddybrando** [comment]
> In my opinion, our project isn't too big and hasn't too many complicated components, so I think `Console` and `Paddings` aren't necessary right now.

I can agree with both.

For **Console** I didn't know yet if we could benefit from it since my overview of the project's components is still limited.

For **Paddings** I can also agree that it's not actually necessary at the moment.

So, I won't configure these two 🙂.

> `Storyshots` and `Screenshot` are both for taking snapshots, what differences do you find and why do you think we should add both?

It's quite tricky but they mean different things by "snapshots":

- **Storyshots** is for [snapshot testing](https://jestjs.io/docs/en/snapshot-testing)
It can help us reuse testing code.
We can also leave this for later when we tackle our testing strategies. What do you think @y4izus?
- **Screenshot** takes literal visual snapshots (screenshots) via Puppeteer and saves these as image files
It can be helpful to have these image assets to easily paste on GitHub issues, slides, Slack, or just to have as quick visual reference.
More of a _nice to have_ than a need.
One caveat is that it requires Puppeteer, and thus requires a headless Chrome, which adds quite some size to the development build.

**Comment by korgan00** [comment]
Very interesting add-ons,
I agree with you; I think we can delay de `storyshots` for later.

**Comment by y4izus** [comment]
Storyshots has also integration [with Puppeteer](https://github.com/storybookjs/storybook/tree/master/addons/storyshots/storyshots-puppeteer). There are differences between this and `Screenshot`?

Anyways, I agree with delaying this add-on and do it when we tackle the testing strategy

**Comment by techtolentino** [comment]
thank you for the breakdown of these add-ons, @eddybrando 

I agree with everything that's been assessed so far.

I'll add a couple of potential nice-to-haves, for consideration... 

- ~[responsive views](https://github.com/vizeat/storybook-addon-responsive-views)~ (disregard, what I was looking for, might already be baked into storybook)
- [themes](https://github.com/tonai/storybook-addon-themes), (low priority)

**Comment by eddybrando** [comment]
> * [themes](https://github.com/tonai/storybook-addon-themes), (low priority)

Thanks @techtolentino!

@korgan00, I believe we could implement this later once we have the theming integrated 🙂. What do you think?

**Reply by korgan00** [reply]
yeah, i think so

**Comment by eddybrando** [comment]
To summarise...

Add-ons we don't need:

- Console
- Links
- Open Component Code
- Paddings
- Screenshot
- Source

Add-ons that may be interesting to revisit later:

- Storyshots
- Themes

Add-ons that we will integrate now:

- Accessibility
- Design (Figma)
